### ISSUE TITLE
[BUG]: Investigate whether the thrust deprecated compiler warnings are still in effect

### ISSUE THREAD

**Issue by miscco** [NVIDIA/cccl#1620]
### Is this a duplicate?

- [X] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this bug and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Type of Bug

Something else

### Component

Thrust

### Describe the bug

We added system header suppression to all our headers. 

However, in `thrust/thrust/detail/config/cpp_dialect.h` we rely on `GCC warning` to relate a deprecated host compiler warning.

We should investigate, whether that warning is still active and otherwise remove the system header suppression from that specific file

### How to Reproduce

Someone would need to build against a very old GCC

### Expected behavior

We should surface a warning

**Comment by SilentSobs** [comment]
Hello,

I would like to work on this issue, but I'm currently stuck.

I tried installing an older version of GCC and G++:
```
g++ --version
g++ (Ubuntu 9.5.0-4ubuntu2) 9.5.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions. There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

However, when I run a test to check for the warning:

```
./test_thrust.sh -cxx g++-9

```

I don't get any output regarding this. Am I doing this correctly?

**Comment by miscco** [comment]
The deprecation is only enabled for gcc-5 or clang-6 and below, so using gcc-9 is expected to pass without a warning

**Comment by SilentSobs** [comment]
![Screenshot from 2024-04-16 03-07-13](https://github.com/NVIDIA/cccl/assets/157403288/2051b043-633c-43ff-b1a2-d97c8adf4fe8)

**Comment by dunga1k58bh** [comment]
@miscco 
Is it necessary to build against very old GCC versions like GCC 5?

I added the following lines to cpp_dialect.h:
```
// TEST WARNING: This should appear with current compilers to prove the point
#  if _CCCL_COMPILER(GCC) || _CCCL_COMPILER(CLANG)
THRUST_COMPILER_DEPRECATION(TEST WARNING - This message proves system header pragma suppresses warnings);
```
Then I built a simple test using the cuda13.0-gcc14 devcontainer.
However, even though there is no condition like # if _CCCL_COMPILER(GCC, <, 7), the warning did not appear:

```
(cccl) coder ➜ ~/cccl $ cat > /tmp/test_warning_with_pragma.cpp << 'EOF'
> #include <thrust/version.h>
> #include <thrust/detail/config/cpp_dialect.h>
> 
> int main() {
>     return 0;
> }
> EOF
(cccl) coder ➜ ~/cccl $ cd /tmp && g++ -std=c++17 -I/home/coder/cccl/thrust -I/home/coder/cccl/libcudacxx/include -c test_warning_with_pragma.cpp 2>&1
(cccl) coder ➜ /tmp $ 
```

After that, I removed the system-header suppression in cpp_dialect.h (and in the .h files it includes), and tried again.
This time, the warning did appear:

```
(cccl) coder ➜ /tmp $ cd /tmp && g++ -std=c++17 -I/home/coder/cccl/thrust -I/home/coder/cccl/libcudacxx/include -c test_warning_with_pragma.cpp 2>&1
In file included from /home/coder/cccl/thrust/thrust/detail/config/config.h:29,
                 from /home/coder/cccl/thrust/thrust/version.h:31,
                 from test_warning_with_pragma.cpp:1:
/home/coder/cccl/thrust/thrust/detail/config/cpp_dialect.h:59:13: warning: Thrust requires at least TEST WARNING - This message proves system header pragma suppresses warnings. Define CCCL_IGNORE_DEPRECATED_COMPILER to suppress this message.
   59 | THRUST_COMPILER_DEPRECATION(TEST WARNING - This message proves system header pragma suppresses warnings);
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
```

So can we conclude that the system-header suppression is preventing the compiler from showing the warning, or do we still need to reproduce the bug using a very old GCC?

**Comment by bernhardmgruber** [comment]
> So can we conclude that the system-header suppression is preventing the compiler from showing the warning

That's important to know, since we emit warnings about the environment in a couple of places (like using MSVC with the old preprocessor). We should audit all places where we emit a warning on purpose and make sure it's not suppressed.

@miscco how should we do this best? I guess we should not just strip the system header macros?

@dunga1k58bh and @SilentSobs thx for checking this!

**Comment by miscco** [comment]
Yeah I believe we should have some tests that build with the system header live and verify that we actually fail with the expected warning.

I would even be fine to ensure that we fail at all
